explainer-rendering.md

```diff
@@ -0,0 +1,110 @@
+Formatted Text - Rendering
+=============
+An approach to rendering multi-line formatted text to an HTML canvas context.
+
+This explainer focuses on rendering [the data model](explainer-datamodel.md) for formatted text.
+For a general overview of the problem space, see the repo's [readme](readme.md). In addition to rendering,
+the data model can [provide metrics](explainer-metrics.md).
+
+---
+## Specifying the layout constraints
+
+## Initial values for CSS properties, esp. when used with the Canvas. Who wins?
+
+## CSS unit value resolution (percentages, px, vh, etc.)
+
+
+
+Next, to render the text into multiple lines, we use the new `fillFormattedText` API
+which takes a max width to wrap the text at.
+
+```js
+context.font = "18pt Arial";
+context.fillFormattedText( formattedText, /*x*/0, /*y*/50, /*wrapWidth*/250 );
+```
+
+This would produce the following output on the canvas:
+
+<img src="explainerresources/Example1.png" alt="Wrapped text rendered in a canvas." align="center"/>
+
+### Bidi Text
+
+No additional work is needed from web developers to support bidi text.
+Implementations will perform bidi analysis on the `CanvasFormattedText`'s text runs
+and create internal bidi runs if necessary. An example demonstrating bidi text follows:
+
+```js
+const context = document.getElementById( "myCanvas" ).getContext( "2d" );
+context.font = "30px Arial";
+let canvasFormattedText = new CanvasFormattedText();
+canvasFormattedText.appendRun( { text: "Sample arabic بمدينة مَايِنْتْس، ألمانيا text." } );
+context.fillFormattedText( canvasFormattedText, /*x*/0, /*y*/30, /*wrapWidth*/350 );
+```
+
+produces the following output on the canvas:
+
+<img src="explainerresources/Example2.png" alt="Wrapped text rendered in a canvas." align="center"/>
+
+### Pre-existing text controls
+
+The other text styles `textAlign`, `textBaseline` on the canvas context control justification
+and baseline alignment of the multiline text relative to provided x/y coordinates.
+
+CSS styles on the canvas element affect text rendering. These CSS properties are
+
+- `line-height` - Specifies height of a line.
+- `direction` - Sets the initial direction for bidi analysis.
+- `word-break` / `word-wrap` - Controls break opportunities for text wrapping.
+
+
+## WebIDL
+
+```webidl
+partial interface CanvasText { 
+  // Perform layout and render FormattedText
+  void fillFormattedText(FormattedText formattedText, 
+                         double x, 
+                         double y, 
+                         double inlineSize, 
+                         optional FormattedTextOptions options); 
```

Comment:
Can we leave this as optional blocksize, rather than introduce the dictionary.

Reply:
Can do!

Reply:
See https://github.com/WICG/canvas-formatted-text/commit/0686b82c39e25fd9692e17f47c278a6afa3b680d